src/whylogs/v2/core/metrics/metric_plugin.py

```diff
@@ -0,0 +1,137 @@
+            raise TypeError("The given plugin does not contain a definition for Python, the plugin types are ({})".format(message.plugin_types))
+        target_class_name = python_plugin.plugin_class_name
+
+        plugin_class = next(
+            filter(
+                lambda plugin_class: MetricPlugin.name_type_match_predicate(target_class_name, plugin_class, enforce_strict_name_matching),
+                MetricPlugin.get_subclasses(),
+            ),
+            None,
+        )
+        if plugin_class is None:
+            raise TypeError(
+                "There is no defined subclass of MetricPlugin matching serialized name {}, the plugin types are: {}".format(
+                    target_class_name, [plugin for plugin in MetricPlugin.get_subclasses()]
+                )
+            )
+        custom_serialization = message.WhichOneof("item")
+        if custom_serialization != "struct":
+            raise TypeError(
+                "The given plugin message for {} does not contain a struct, if using serialized_bytes override from_protobuf() to avoid this error.".format(
+                    target_class_name
+                )
+            )
+        plugin_field = json_format.MessageToDict(message.struct)
+        metric_plugin = plugin_class(**plugin_field)
+        metric_plugin.name = message.name
+        return metric_plugin
+
+    def to_protobuf(
+        self,
+    ) -> PluginMessage:
+        # Authors can override, in base implementation here we support python using full module.qualname
+        # Note: there is no version checking here with respect to changes in your plugin.
+        types = [
+            PluginType(
+                plugin_class_name=MetricPlugin.get_name_from_type(self.__class__),
+                language=PluginType.Language.PYTHON,
+            )
+        ]
+        plugin_message = PluginMessage(
+            name=self.name,
+            plugin_types=types,
+        )
+        # The protobuf struct will serialize the field names as well as the values, so plugin
+        # authors might get better efficiency if you override and serialize to bytes instead.
+        plugin_fields = asdict(self)
+        # we don't want to duplicate the name field in the serialized fields as
+        # we already have a 'name' property on PluginMessage
+        plugin_fields.pop("name", None)
+        plugin_message.struct.update(plugin_fields)
+        return plugin_message
+
+    @abstractmethod
+    def track(self, data):
+        pass
+
+    @abstractmethod
+    def merge(self, other: "MetricPlugin"):
+        pass
+
+    def to_summary(self):
+        return self.serialize()
+
+    @staticmethod
+    def deserialize(data: bytes, enforce_strict_name_matching: bool = False) -> "MetricPlugin":
+        message = PluginMessage.FromString(data)
+        return MetricPlugin.from_protobuf(message, enforce_strict_name_matching)
+
+    def to_json(self) -> dict:
+        return asdict(self)
```

Comment:
You might want to distinguish top-level fields like name and have subclass params as 'params' to avoid clashes and do the same for deserialization.

```python
return {'cls_name': name() , 'params': asdict(self)}
```

Reply:
Good call, I'll fix that up to handle the top level fields better. Thanks for taking a look!

Reply:
Also just as an option try exploring https://lidatong.github.io/dataclasses-json/  for custom serializer and deserializers

Reply:
dataclasses_json handles nested classes and private fields better, thanks. I created a dev release from this to try out:
https://pypi.org/project/whylogs/0.6.4.dev0/

You can use try the example notebook: https://github.com/whylabs/whylogs/blob/dev/jamie/v2_experimental/examples/custom_metrics.ipynb